### DISCUSSION TITLE
Connecting to nREPL is slow?

### DISCUSSION THREAD

**Discussion by liquidz** [liquidz/vim-iced#320] [General]
It took less than one second in my environment, but it seems to take some people a few seconds.

**Comment by liquidz** [comment]
@Deraen Is it still slow if you specify the nREPL port number?
e.g. `:IcedConnect <nREPL port>`

@NoahTheDuke Feel free to comment if you have any question about this topic.

**Reply by Deraen** [reply]
Yes, still slow when connecting specified port.

I recall that Cider-nrepl might be sending lots of data when the connection (session?) is created? Or maybe that was one specific middleware that is already disabled? Could be something related to that, or should all the be happening async?

I've recently mostly worked on a project with about 100 namespaces and 10k lines, but occasionally working with projects with 50k lines.

Connecting to opensource project with single ns and 300 lines is definitely faster, probably under 1 second (still noticeable if you are already running next commands), while connecting to large project might take even 5 seconds.

**Reply by liquidz** [reply]
Hmm, vim-iced does not use `track-state` middleware, so it should not occur large amount of data sending on connecting.
c.f. https://github.com/liquidz/vim-iced/issues/170

There are two sync processe when connecting for now.
- [Cloning session](https://github.com/liquidz/vim-iced/blob/3.0.1/autoload/iced/nrepl.vim#L497)
- [Fetching ns name](https://github.com/liquidz/vim-iced/blob/3.0.1/autoload/iced/nrepl.vim#L434)
  - Just evaluating `(ns-name *ns*)`
 
The former is sync because it is used to get a session to communicate with nREPL, but the latter could be made async I think.

And the latter seems to be slow when the first evaluation after starting nREPL server.
e.g. [antq](https://github.com/liquidz/antq)
```sh
### The first connection after starting nREPL server
DEBUG s:detect_shadow_cljs_nrepl_port: 6.03e-4
DEBUG s:detect_port_from_nrepl_port_file: 0.001389
DEBUG iced#nrepl#clone(SYNC): 0.021669
DEBUG iced#nrepl#ns#name_by_var(SYNC): 1.05377.         <----- 1sec
DEBUG describe(iced#nrepl#is_supported_op): 0.018793
DEBUG iced#nrepl#set_iced_nrepl_enabled: 1.3e-5
DEBUG s:warm_up: 0.121358
Connected.
DEBUG connected: 1.58155

### The second connection after starting nREPL server
DEBUG s:detect_shadow_cljs_nrepl_port: 5.51e-4
DEBUG s:detect_port_from_nrepl_port_file: 0.001164
DEBUG iced#nrepl#clone(SYNC): 0.031086
DEBUG iced#nrepl#ns#name_by_var(SYNC): 0.01399
DEBUG describe(iced#nrepl#is_supported_op): 0.01063
DEBUG iced#nrepl#set_iced_nrepl_enabled: 1.0e-5
DEBUG s:warm_up: 0.11027
Connected.
DEBUG connected: 0.505123
```

I created a branch to output the debug log that measures time.
https://github.com/liquidz/vim-iced/tree/connection_time

**Reply by Deraen** [reply]
Doesn't seem like the slow part is name_by_var for me, at least in case where I have the project system already running, and all namespaces are compiled:

```
DEBUG iced#nrepl#clone(SYNC): 0.010884
DEBUG iced#nrepl#ns#name_by_var(SYNC): 0.006292
DEBUG describe(iced#nrepl#is_supported_op): 0.003864
DEBUG iced#nrepl#set_iced_nrepl_enabled: 1.4716e-5
DEBUG s:warm_up: 5.586125
Connected.
DEBUG connected: 8.51458
```

When starting up a new REPL, if namespaces need to be compiled, that could take some time.

**Reply by Deraen** [reply]
It is the `if iced#nrepl#is_supported_op('init-debugger')` part which takes several seconds in my setup. Specifically the `sleep 100ms` part takes a lot longer than just 100 milliseconds?

**Reply by Deraen** [reply]
But even after commenting that part out, and the debug prints showing "connected" message, my cursor is still frozen for a few seconds after that.

**Reply by liquidz** [reply]
Thanks!

> It is the if iced#nrepl#is_supported_op('init-debugger') part which takes several seconds in my setup. Specifically the sleep 100ms part takes a lot longer than just 100 milliseconds?

Oh sorry, `is_supported_op` had a sync process.
https://github.com/liquidz/vim-iced/blob/3.0.1/autoload/iced/nrepl.vim#L661
But it's just sending a `describe` op to nREPL and its response should no be so large... I have no idea why the part takes time.

> But even after commenting that part out, and the debug prints showing "connected" message, my cursor is still frozen for a few seconds after that.

It's possible that some huge data is received and frozen in parsing.
You can confirm all input/output between nREPL with `let g:iced#debug = v:true` (but it can be very noisy)
All input/output are displayed in `:messages`
